clang_build/project.py

```diff
@@ -288,35 +351,80 @@ def _fill_dependency_tree(self):
 
             target.config["dependencies"] = dependency_objs
 
-        # Write dotfile with full dependency graph
+        # Create new dotfile
         if create_dotfile:
             _nx.drawing.nx_pydot.write_dot(self._project_tree, str(self._environment.build_directory / 'dependencies.dot'))
+        self._check_for_circular_dependencies()
 
-    def _load_config(self):
-        """Try to find a toml file and return a config.
+        # Write dotfile with full dependency graph
+        if not self.parent:
+            self._check_for_circular_dependencies()
 
-        This helper function is part of the initialisation of a project.
-        It searches for a toml file and (if it finds one) loads it. Otherwise
-        it generates a default config.
+        # Update the `only_target` flag of the targets in this project
+        n_targets = len(self._current_targets)
+        only_target = n_targets == 1 and len(self.config.get("subprojects", [])) == 0
+
+        for target in self._current_targets:
+            target.only_target = only_target
+
+
+    def _parse_subprojects(self):
+        """Generate all subprojects recursively.
+
+        Recursively walks through all subprojects and parses
+        them.
+
+        Returns
+        -------
+        list
+            List of Project objects.
         """
-        toml_file = self.directory / "clang-build.toml"
+        subproject_list = []
+        for directory in self.config.get("subprojects", []):
+            subproject_dir = self._directory / directory
+            if (subproject_dir / "clang-build.py").exists():
```

Comment:
Your `from_directory` covers both cases: toml and py. Why do you do the checks here again?

Reply:
I do it like this in order to have better logging messages, because I did not find the log message `f"Using python project file \"{py_file}\""` printed from the subproject to be very clear.

Reply:
I don't think we have to change the code structure for better logging. This seems to have the wrong priorities.

Reply:
You are right. I was able to find a much better way: simply check if `parent` exists and, if so, slightly change the log message.